File: Java/817. Linked List Components.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int numComponents(ListNode head, int[] nums) {
        HashMap<Integer, Integer> mp = new HashMap<Integer, Integer>();
        for(int i: nums){
            mp.put(i,1);
        }
        int res = 0;
        while(head!=null){
            if(mp.get(head.val)!=null){
                while(head.next!=null && mp.get(head.next.val)!=null){
                    head = head.next;
                }
                res++;
            }
            head = head.next;
        }
        return res;
    }
}

File: C++/817. Linked List Components.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int numComponents(ListNode* head, vector<int>& nums) {
        map<int,int> mp;
        for(int i: nums){
            mp[i]++;
        }
        int res = 0;
        while(head!=NULL){
            if(mp[head->val]!=NULL){
                while(head->next!=NULL && mp[head->next->val]!=NULL){
                    head = head->next;
                }
                res++;
            }
            head = head->next;
        }
        return res;
    }
};
